packages/protocol/contracts/common/TaikoProxy.sol

```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+//
+//   Email: security@taiko.xyz
+//   Website: https://taiko.xyz
+//   GitHub: https://github.com/taikoxyz
+//   Discord: https://discord.gg/taikoxyz
+//   Twitter: https://twitter.com/taikoxyz
+//   Blog: https://mirror.xyz/labs.taiko.eth
+//   Youtube: https://www.youtube.com/@taikoxyz
+
+pragma solidity 0.8.24;
+
+import "lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import "lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+
+/// @title TaikoProxy
+/// @dev An ERC1967Proxy that allows for reading its implementation address.
+contract TaikoProxy is ERC1967Proxy {
+    constructor(address _logic, bytes memory _data) payable ERC1967Proxy(_logic, _data) { }
+
+    function implementation() public view returns (address) {
```

Comment:
One thing to be careful for now though is that a call to `implementation()` will always end up in the proxy contract, so any function called `implementation()` in the implementation contract cannot be reached.

Perhaps giving it more of unique name like `getProxyImplementation()` would be better because a conflict is already much more unlikely.